providers/kubeconfig/provider.go

```diff
@@ -0,0 +1,593 @@
+	DefaultKubeconfigSecretKey = "kubeconfig"
+)
+
+var _ multicluster.Provider = &Provider{}
+
+// New creates a new Kubeconfig Provider.
+func New(opts Options) *Provider {
+	// Set defaults
+	if opts.KubeconfigLabel == "" {
+		opts.KubeconfigLabel = DefaultKubeconfigSecretLabel
+	}
+	if opts.KubeconfigKey == "" {
+		opts.KubeconfigKey = DefaultKubeconfigSecretKey
+	}
+	if opts.ConnectionTimeout == 0 {
+		opts.ConnectionTimeout = 10 * time.Second
+	}
+	if opts.CacheSyncTimeout == 0 {
+		opts.CacheSyncTimeout = 30 * time.Second
+	}
+	if opts.KubeconfigPath == "" {
+		opts.KubeconfigPath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	}
+
+	return &Provider{
+		opts:        opts,
+		log:         log.Log.WithName("kubeconfig-provider"),
+		client:      nil, // Will be set in Run
+		clusters:    map[string]cluster.Cluster{},
+		cancelFns:   map[string]context.CancelFunc{},
+		seenHashes:  map[string]string{},
+		readySignal: make(chan struct{}),
+	}
+}
+
+// Options are the options for the Kubeconfig Provider.
+type Options struct {
+	// Namespace to watch for kubeconfig secrets
+	Namespace string
+
+	// Label key to identify kubeconfig secrets
+	KubeconfigLabel string
+
+	// Key in the secret data that contains the kubeconfig
+	KubeconfigKey string
+
+	// ConnectionTimeout is the timeout for connecting to a cluster
+	ConnectionTimeout time.Duration
+
+	// CacheSyncTimeout is the timeout for waiting for the cache to sync
+	CacheSyncTimeout time.Duration
+
+	// KubeconfigPath is the path to the kubeconfig file to use for development/testing
+	// If not set, will use the default ~/.kube/config
+	KubeconfigPath string
+}
+
+type index struct {
+	object       client.Object
+	field        string
+	extractValue client.IndexerFunc
+}
+
+// Provider is a cluster provider that watches for secrets containing kubeconfig data
+// and engages clusters based on those kubeconfigs.
+type Provider struct {
+	opts        Options
+	log         logr.Logger
+	client      client.Client
+	lock        sync.RWMutex
+	clusters    map[string]cluster.Cluster
+	cancelFns   map[string]context.CancelFunc
+	indexers    []index
+	seenHashes  map[string]string // tracks resource versions
+	readySignal chan struct{}     // Signal when provider is ready to start
+	readyOnce   sync.Once         // Ensure we only signal once
+}
+
+// IsReady returns a channel that will be closed when the provider is ready to start
+func (p *Provider) IsReady() <-chan struct{} {
+	return p.readySignal
+}
+
+// Get returns the cluster with the given name, if it is known.
+func (p *Provider) Get(ctx context.Context, clusterName string) (cluster.Cluster, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if cl, ok := p.clusters[clusterName]; ok {
+		return cl, nil
+	}
+
+	return nil, fmt.Errorf("cluster %s not found", clusterName)
+}
+
+// Run starts the provider and blocks, watching for kubeconfig secrets.
+func (p *Provider) Run(ctx context.Context, mgr mcmanager.Manager) error {
```

Comment:
I'll be honest, this func looks much more complex than I think it has to be. Did you take a look at how https://github.com/kubernetes-sigs/multicluster-runtime/blob/main/providers/namespace/provider.go does it? Working with informers based on a `cluster.Cluster` and its cache seems much more straightforward.

Reply:
I've simplified this greatly